### ISSUE TITLE
[JAVA] isJsonMime should support vendor specific JSON MIME type

### ISSUE THREAD

**Issue by davidharrigan** [swagger-api/swagger-codegen#5172]
##### Description

We support APIs that return JSON encoded responses with vendor specific mime types.  These MIME types look something like:
```
application/vnd.company+json
```
Which is a valid MIME-type. A response with the above mime-type as its content-type indicates that the response is JSON encoded.  When the swagger-client attempts to deserialize such response, it throws an exception due to a method responsible for determining whether the given MIME is a JSON MIME.  The method is [`isJsonMime`](https://github.com/swagger-api/swagger-codegen/blob/3c1a8daaf920af7a5eecf17f3fb157f083ca6366/modules/swagger-codegen/src/main/resources/Java/libraries/okhttp-gson/ApiClient.mustache#L728) and it checks whether the mime type matches the following regex:
```
(?i)application\\/json(;.*)?
```

This results in the following error:
```
Content type "application/vnd.company+json" is not supported for type: class com.somemodel Model
```

Not tested, but I think the above regex should be updated to something like:

```
(?i)application\\/((\w\.)*\w\+)?json(;.*)?"
```
so that MIME types like `application/vnd.company+json` are recognized as a valid JSON MIME.

I think it would be valid to have some additional tests [here](https://github.com/swagger-api/swagger-codegen/blob/1dd9ee39af6c4e2f9c016a94d16ca47a88a9cb0d/samples/client/petstore/java/okhttp-gson/src/test/java/io/swagger/client/ApiClientTest.java#L88) 
that checks for MIME types such as the one mentioned above.  

##### Swagger-codegen version

`v2.2.1`

##### Swagger declaration file content or url

N/A

##### Command line used for generation

```bash
java -jar /opt/swagger-codegen/modules/swagger-codegen-cli/target/swagger-codegen-cli.jar generate -i /mnt/swagger.yaml -l java -o /mnt/clients/java -c /mnt/clients/configs/java.json
```

##### Steps to reproduce

- Generate a client for some API
- Have the API return a vendor JSON MIME type as the content header
- use the client to retrieve from the endpoint which produces a vendor JSON MIME type
- watch it scream

##### Related issues

This is a simpler fix, but I believe https://github.com/swagger-api/swagger-codegen/issues/3870 is related.

##### Suggest a Fix

Mentioned in the description.

**Comment by gibson042** [comment]
`isJsonMime` should accept anything matching `application/json` or [ending with](https://tools.ietf.org/html/rfc6838#section-4.2) `+json` (cf. [RFC 6839](https://tools.ietf.org/html/rfc6839#section-3.1)), in both cases case-insensitive and after stripping off everything at or after the first semicolon and any (remaining) leading/trailing whitespace (cf. [RFC 7231](https://tools.ietf.org/html/rfc7231#section-3.1.1.1) and [RFC 7230](https://tools.ietf.org/html/rfc7230#section-3.2)). In code, something like
```java
mime.matches("(?i)^(application/json|[^;/ \t]+/[^;/ \t]+[+]json)[ \t]*(;.*)?$")
```
http://ideone.com/MCsVFO

**Comment by wing328** [comment]
Closed via #5189
